### PR TITLE
fix: only optimize annoymous iife

### LINKED ISSUE #5484: Treeshaking breaks acorn-walk "simple"
### Rollup Version

14.6.2

### Operating System (or Browser)

osx

### Node Version (if applicable)

20

### Link To Reproduction

https://rollupjs.org/repl/?version=4.16.2&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMHNpbXBsZSUyMCU3RCUyMGZyb20lMjAnLiUyRndhbGtlciclM0IlNUNuJTVDbnNpbXBsZSglN0IlMjB0eXBlJTNBJTIwJTVDJTIySWRlbnRpZmllciU1QyUyMiUyMCU3RCUyQyUyMCU3QiU1Q24lMjAlMjBFeHByZXNzaW9uKG5vZGUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKCU1QyUyMm5vZGUlMjBmb3VuZCU1QyUyMiklNUNuJTIwJTIwJTdEJTVDbiU3RCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSUyQyUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTdEJTJDJTdCJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMGh0dHBzJTNBJTJGJTJGZ2l0aHViLmNvbSUyRmFjb3JuanMlMkZhY29ybiUyRmJsb2IlMkY4ODIzNGY1MjlmOWZkYzNhYTI3MzZhNjU3NWEzNzc4ZTAzMTExMTBkJTJGYWNvcm4td2FsayUyRnNyYyUyRmluZGV4LmpzJTIzTDE5LUwyNiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHNpbXBsZShub2RlJTJDJTIwdmlzaXRvcnMlMkMlMjBiYXNlVmlzaXRvciUyQyUyMHN0YXRlJTJDJTIwb3ZlcnJpZGUpJTIwJTdCJTVDbiUyMCUyMGlmJTIwKCFiYXNlVmlzaXRvciklMjBiYXNlVmlzaXRvciUyMCUzRCUyMGJhc2UlNUNuJTIwJTIwJTNCKGZ1bmN0aW9uJTIwYyhub2RlJTJDJTIwb3ZlcnJpZGUpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGxldCUyMHR5cGUlMjAlM0QlMjBvdmVycmlkZSUyMCU3QyU3QyUyMG5vZGUudHlwZSU1Q24lMjAlMjAlMjAlMjBiYXNlVmlzaXRvciU1QnR5cGUlNUQobm9kZSUyQyUyMGMpJTVDbiUyMCUyMCUyMCUyMGlmJTIwKHZpc2l0b3JzJTVCdHlwZSU1RCklMjB2aXNpdG9ycyU1QnR5cGUlNUQobm9kZSklNUNuJTIwJTIwJTdEKShub2RlJTJDJTIwb3ZlcnJpZGUpJTVDbiU3RCU1Q24lNUNuZXhwb3J0JTIwY29uc3QlMjBiYXNlVmlzaXRvciUyMCUzRCUyMCU3QiU1Q24lMjAlMjBJZGVudGlmaWVyKG5vZGUlMkMlMjBjKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjKG5vZGUlMkMlMjAlNUMlMjJFeHByZXNzaW9uJTVDJTIyKSU1Q24lMjAlMjAlN0QlNUNuJTdEJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMndhbGtlci5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=

### Expected Behaviour

acorn-walk simple visitor methods should be called after compilation.

In the provided repl, the node of type Identifier should trigger a recursive call to "c" with override = "Expression" and that should `if (visitors[type]) visitors[type](node)` -> console.log("node found")

If the `override` parameter is assumed undefined with treeshaking (even "safest" exhibits this issue), then the behavior is broken.

### Actual Behaviour

acorn-walk simple visitor methods are not called because "override" is erroneously optimized out.

### PR DESCRIPTION
**Pull request by liuly0322**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

- resolves #5484

### Description

If a IIFE has a name, it may lose track. In this case we probably should not consider to optimize them.

### CODE REVIEW

File: src/ast/nodes/shared/FunctionBase.ts [hunk 1]
```diff
@@ -317,10 +316,7 @@ export default abstract class FunctionBase extends NodeBase {
 
 	private functionParametersOptimized = false;
 	include(context: InclusionContext, includeChildrenRecursively: IncludeChildren): void {
-		const isIIFE =
-			this.parent.type === NodeType.CallExpression &&
-			(this.parent as CallExpression).callee === this;
-		const shoulOptimizeFunctionParameters = isIIFE || this.onlyFunctionCallUsed();
+		const shoulOptimizeFunctionParameters = this.onlyFunctionCallUsed();
```
Comment (sparecycles) [inline src/ast/nodes/shared/FunctionBase.ts:319]: Hey, I can contribute a spelling fix, at least 😉 

```suggestion
		const shouldOptimizeFunctionParameters = this.onlyFunctionCallUsed();
```

File: src/ast/nodes/FunctionExpression.ts [hunk 2]
```diff
@@ -25,6 +26,14 @@ export default class FunctionExpression extends FunctionNode {
 		return super.parseNode(esTreeNode);
 	}
 
+	protected onlyFunctionCallUsed(): boolean {
+		const isIIFE =
+			this.parent.type === NodeType.CallExpression &&
+			(this.parent as CallExpression).callee === this &&
+			!this.id;
+		return isIIFE || super.onlyFunctionCallUsed();
+	}
```
Comment (sparecycles) [inline src/ast/nodes/FunctionExpression.ts:35]: so it looks like the functional change here is making `isIIFE` compute to false for named functions?

(slightly better would be to check that the function value is referenced anywhere, I don't know how easy/quick that is, right?)

**Review by sparecycles — COMMENTED** [review]
lgtm, I think? (Don't take my word for it, I'm new here!)

File: src/ast/nodes/FunctionExpression.ts (hunk 2 above)
Comment (liuly0322) [reply src/ast/nodes/FunctionExpression.ts]: Yeah we can check it carefully to support to  no reference or recursion.

**Review by lukastaegert — APPROVED** [review]
Good work!

**Pull request merged**